### ISSUE TITLE
Running Scrapy using CrawlRunner and restarting reactor - Recommendation from Scrapy Docs

### ISSUE THREAD

**Issue by chipzzz** [scrapy/scrapy#5026]
I am using the following code as suggested from [Scrapy doc](https://docs.scrapy.org/en/latest/topics/practices.html#run-scrapy-from-a-script),  when running spiders within the same reactor.  However, with this implemented I still get `twisted.internet.error.ReactorNotRestartable`

Did anyone get this piece of code to successfully run? For now, I'm adding `os.execl(sys.executable, sys.executable, *sys.argv)` to the end, to kill the process manually, it works and is good because I don't want to run multi processes as others suggest as I'm running this within a container ECS) 

```
from twisted.internet import reactor
import scrapy
from scrapy.crawler import CrawlerRunner
from scrapy.utils.log import configure_logging

class MySpider(scrapy.Spider):
    # Your spider definition
    ...

configure_logging({'LOG_FORMAT': '%(levelname)s: %(message)s'})
runner = CrawlerRunner()

d = runner.crawl(MySpider)
d.addBoth(lambda _: reactor.stop())
reactor.run() # the script will block here until the crawling is finished
```

My actual code, I'm starting my crawler from a separate execution script does some extra things, like checks s3 etc.

```
while True:
    read_json = check_s3()
    configure_logging({'LOG_FORMAT': '%(levelname)s: %(message)s'})
    runner = CrawlerRunner(get_project_settings())
    print("STARTING_SCRAPE ", read_json[1])
    d = runner.crawl(MenuSpider, menu_sites=read_json[0], lua_script='crawlera_script.lua')
    d.addBoth(lambda _: reactor.stop())
    reactor.run()
    print('Sleeping for 30 seconds')
    time.sleep(30)
    #os.execl(sys.executable, sys.executable, *sys.argv)
```

**Comment by Gallaecio** [comment]
Do you want to make this a request to enhance the documentation to cover a recommendation on how to run a crawler in a loop? If so, that’s OK, but I think the solution will be through multiprocessing or similar, which [does not seem to be what you are looking for](https://github.com/scrapy/scrapy/issues/2941#issuecomment-790307550).

In a container environment, the default recommendation would probably be to let the container run the crawler once, and let the container itself be restarted, which should not cause the reactor-not-restartable issue.

If none of those approaches work for you, I’m not sure what to tell you.

I find your actual code a bit confusing. Why the sleep? Does the `while True` run more than once? Isn’t `os.execl(sys.executable, sys.executable, *sys.argv)` a convoluted way to exit the loop?

If you are reporting a bug in Scrapy, please provide a [minimal, reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) instead of a fragment of your actual code. If you want help to get your code working the way you want, please, use [StackOverflow](https://stackoverflow.com/tags/scrapy) (see [Getting Help](https://docs.scrapy.org/en/latest/#getting-help)).

**Comment by wRAR** [comment]
The first code snippet shouldn't have this problem, are you saying it does? Putting it into a loop indeed won't work.

**Comment by wRAR** [comment]
The first code snippet shouldn't have this problem, are you saying it does? Putting it into a loop indeed won't work.

**Comment by wRAR** [comment]
Or, if you want to run several CrawlerRunner instances, you need to only run them in the loop and start/stop the reactor only once (though I admit I'm not sure how should the code look like, at least when not using inlineCallbacks).

**Comment by chipzzz** [comment]
@Gallaecio ,  Yes I am trying to make an improvement to the docs, as I am utilizing what the docs recommend when you want to execute spiders continually and not have problems with the reactor. However, when utilizing the code provided by the docs it doesn't seem to work,  I am assuming that this `d.addBoth(lambda _: reactor.stop())` doesn't work thus reactor doesn't stop, thus getting that error when trying to run reactor again.

Can you elaborate on what this actually does, I can perhaps troubleshoot further `d.addBoth(lambda _: reactor.stop())`.  I am unclear on how this grabs the signal that a crawl has completed, or if it even does at all.

My intention is this: 

A spider that is continually running grabbing fresh urls from s3 bucket , that's why I have the `while True:` loop.

> In a container environment, the default recommendation would probably be to let the container run the crawler once, and let the container itself be restarted, which should not cause the reactor-not-restartable issue.

I don't agree with that, in a perfect world containers shouldn't restart when they don't have to, they restart if they fail. That's like redeploying the entire solution each time, it slows things down. What is preferred is something that is constantly running like a service, so what's I'm trying to imitate.

I have the sleep, just for some time between spider crawls, mostly for testing. Will greatly reduce that for production. `os.execl(sys.executable, sys.executable, *sys.argv)` kills the process after a crawl is completed, I am using this to kill the reactor process just as you suggested with restarting the container, which would accomplish the same thing.

**Comment by chipzzz** [comment]
@wRAR, will try running without the while loop, but what makes it not work? It's simply a set of tasks in a loop, which I need because I would like the spider to keep running.

1. call crawler
2. stop reactor (say 2nd time around and the reactor from the previous loop is still running)
3. run crawl

repeat...

I am NOT trying to run multiple crawls at the same time, I am just trying to run crawler after crawler.

**Comment by wRAR** [comment]
>  but what makes it not work?

You cannot run the reactor for the second time. That's what the error says.

>  I am assuming that this d.addBoth(lambda _: reactor.stop()) doesn't work thus reactor doesn't stop, thus getting that error when trying to run reactor again.

It stops, otherwise the code after `reactor.run()` wouldn't run.

> I am unclear on how this grabs the signal that a crawl has completed, or if it even does at all.

The Deferred returned from `runner.crawl` is fired when that crawl finishes.

> A spider that is continually running grabbing fresh urls from s3 bucket , that's why I have the while True: loop.

You can make that with `spider_idle` instead of multiple sequential crawls.

**Comment by chipzzz** [comment]
@wRAR , thanks I'll look into spider_idle... 

> You cannot run the reactor for the second time. That's what the error says.

> > I am assuming that this d.addBoth(lambda _: reactor.stop()) doesn't work thus reactor doesn't stop, thus getting that error when trying to run reactor again.

> It stops, otherwise the code after reactor.run() wouldn't run.

Well isn't the whole point of `d.addBoth(lambda _: reactor.stop())` for the reactor to stop running so that it can be started again without that error?

**Comment by wRAR** [comment]
The whole point of d.addBoth(lambda _: reactor.stop()) for the reactor to stop running. It cannot be started again.

**Comment by chipzzz** [comment]
Sounds good, perhaps I misinterpreted the docs against my use case. - Thanks for the feedback.

**Comment by Gallaecio** [comment]
Mind that the Scrapy documentation does not cover how to run a spider continuously, but there are many examples out there on how to do it. There’s usually 2 different approaches, [the approach at the spider level](https://stackoverflow.com/a/41447672/939364) that @wRAR mentions and which is most often used for your use case, and [the approach at the crawler runner level](https://stackoverflow.com/a/44230214/939364).

**Comment by chipzzz** [comment]
@Gallaecio , thanks for linking those I'll try them out!
